Add Hathora STT and TTS services

mikeseese opened the pull request:
#### Please describe the changes in your PR. If it is addressing an issue, please reference that as well.

This PR adds a new `hathora` service to provide STT and TTS services [hosted by Hathora](https://models.hathora.dev) via our unified API. We currently support Parakeet TDT STT and Kokoro TTS models with more coming soon.

This PR includes an example which also uses the `OpenAILLMService` with a Hathora-hosted Qwen3 model.

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,107 @@
+        self._start_time = start_time
+        self._end_time = end_time
+
+    def can_generate_metrics(self) -> bool:
```
Public methods require docstrings:
```
        """Check if this service can generate metrics.

        Returns:
            True, as Hathora STT service supports metrics generation.
        """
```

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,107 @@
+from pipecat.utils.time import time_now_iso8601
+
+class ParakeetTDTSTTService(SegmentedSTTService):
+    """Parakeet TDT is a multilingual automatic speech recognition model
```
Nit: Docstring linting doesn't allow to wrap the first line in a docstring🤷 

```suggestion
    """Parakeet TDT is a multilingual automatic speech recognition model
```

markbackman commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,229 @@
+    def can_generate_metrics(self) -> bool:
+        return True
+
+    async def run_tts(self, text: str):
```
Same as above, re: docstring.

markbackman replied:
Also, you will want to add tracing to this function, which can be done by decorating the `run_tts()` function:
```suggestion
    @traced_tts
    async def run_tts(self, text: str):
```

Import is:
```
from pipecat.utils.tracing.service_decorators import traced_tts
```

markbackman commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,229 @@
+            await self.stop_processing_metrics()
+            yield TTSStoppedFrame()
+
+class ChatterboxTTSService(TTSService):
```
Same comments about docstrings and `@traced_tts` here, too.

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,107 @@
+                    )
+
+        except Exception as e:
+            logger.error(f"Hathora error: {e}")
```
```suggestion
```

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,107 @@
+
+    def can_generate_metrics(self) -> bool:
+        return True
+
```
Import is:
```
from pipecat.utils.tracing.service_decorators import traced_stt
```

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,107 @@
+                if text:  # Only yield non-empty text
+                    await self.stop_ttfb_metrics()
+                    await self.stop_processing_metrics()
+                    logger.debug(f"Transcription: [{text}]")
```
Call `await self._handle_transcription(transcript, True, "en")` (or similar)

markbackman reviewed (commented):
This is on the right track! I left a bunch of small comments, but nothing major.

A few other things to tend to:
1. Add `HATHORA_API_KEY` to https://github.com/pipecat-ai/pipecat/blob/main/env.example
2. Add the new Hathora services to the README: https://github.com/pipecat-ai/pipecat/blob/main/README.md#-available-services
3. Add a changelog fragment, which you can learn more about here: https://github.com/pipecat-ai/pipecat/blob/main/CONTRIBUTING.md#changelog-entries
4. Make sure to lint your code (scripts/fix-ruff.sh). You can also install the pre-commit hook: `uv run pre-commit install`

The only other question I had was about the service classes. Are these specific to Hathora? I'm assuming so. Also, does each model have a different API or are there common types? Depending on that answer, you might want a:
- `HathoraKokoroTTSService`

Or, if the Kokoro implementation is generic and it should be `KokoroTTSService`, then I'd recommend organizing under `pipecat.services.kokoro.tts`. Then, you can add Hathora as a base url.

I'm looking to make sure these won't be one off classes and will either have extensibility for other base_urls or use for other models. Happy to chat about that point in Slack (or here).

markbackman commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,144 @@
+        model: str,
+        voice: Optional[str] = None,
+        speed: Optional[float] = None,
+        model_config: Optional[list[ConfigOption]] = None,
```
What is this used for? Maybe add to `InputParams` too? Depends on whether these are required values or optional configuration settings.

mikeseese replied:
These are optional config values if the underlying model supports it; it's a passthrough to the Hathora API. Based on that, I'll add them to `InputParams`

markbackman commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,144 @@
+        self._model_config = model_config
+        self._base_url = base_url
+        self._api_key = api_key or os.getenv("HATHORA_API_KEY")
+
```
Also, set the class as being able to generate metrics:
```
def can_generate_metrics(self) -> bool:
        """Check if this service can generate processing metrics.

        Returns:
            True, as Cartesia HTTP service supports metrics generation.
        """
        return True
```

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,131 @@
+        """Handle a transcription result with tracing."""
+        pass
+
+    @traced_stt
```
Only trace your `_handle_transcription()` method.

```suggestion
```

markbackman commented on examples/foundational/07ag-interruptible-hathora.py:
```diff
@@ -0,0 +1,133 @@
+        pipeline,
+        params=PipelineParams(
+            enable_metrics=True,
+            enable_usage_metrics=True,
```
Add a new item, `turn_start_strategies`, to `PipelineParams`:
```
    task = PipelineTask(
        pipeline,
        params=PipelineParams(
            enable_metrics=True,
            enable_usage_metrics=True,
            turn_start_strategies=TurnStartStrategies(
                bot=[TurnAnalyzerBotTurnStartStrategy(turn_analyzer=LocalSmartTurnAnalyzerV3())]
            ),
        ),
        idle_timeout_secs=runner_args.pipeline_idle_timeout_secs,
    )
```

markbackman reviewed (commented):
I like these new classes much better! This will be a more extensible approach and will be developer friendly :)

I've made some comments to help you keep going. In testing, I no longer can run the example. Could this be an issue with my account's access?

mikeseese commented:
Thanks Mark! I'll double check to make sure I didn't break something. I'm off for the holidays, but I'll be back at this on the 5th. Happy Holidays!

mikeseese commented on examples/foundational/07ag-interruptible-hathora.py:
```diff
@@ -0,0 +1,133 @@
+        audio_in_enabled=True,
+        audio_out_enabled=True,
+        vad_analyzer=SileroVADAnalyzer(params=VADParams(stop_secs=0.2)),
+        turn_analyzer=LocalSmartTurnAnalyzerV3(params=SmartTurnParams()),
```
```suggestion
```

mikeseese commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,144 @@
+    [Documentation](https://models.hathora.dev)
+    """
+
+    def __init__(
```
I didn't think adding an unused `sample_rate: Optional[int] = None,` param would be good DevX? They would be inputting one thing and consistently not getting the result they were expecting.

markbackman replied:
That's fine. We can add the arg as sample_rate is controllable.

mikeseese commented:
@markbackman With the exceptions of the unresolved comments above, I believe I addressed the other concerns.

The example will not work as-is because the default API endpoints for base_url are not live yet. You'll need to make the below change to use the dev endpoint:

``` diff
    stt = HathoraSTTService(
        model="nvidia-parakeet-tdt-0.6b-v3",
+        params=HathoraSTTService.InputParams(
+            base_url="https://model-marketplace-api-dev.fly.dev/inference/v1/stt",
+        ),
    )

    tts = HathoraTTSService(
        model="hexgrad-kokoro-82m",
+        params=HathoraTTSService.InputParams(
+            base_url="https://model-marketplace-api-dev.fly.dev/inference/v1/tts",
+        ),
    )
```

I moved `base_url` to the `InputParams` because it should very rarely need to be changed by end users, and would likely only be changed by Hathora devs to test new staging endpoints.

markbackman commented on src/pipecat/services/hathora/stt.py:
```diff
@@ -0,0 +1,155 @@
+
+        language: Optional[str] = None
+        config: Optional[list[ConfigOption]] = None
+        base_url: str = "https://api.models.hathora.dev/inference/v1/stt",
```
I recommend moving the `base_url` to `__init__`, which are generally initialization time items.

markbackman commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,169 @@
+
+        speed: Optional[float] = None
+        config: Optional[list[ConfigOption]] = None
+        base_url: str = "https://api.models.hathora.dev/inference/v1/tts",
```
Same here.

markbackman commented on src/pipecat/services/hathora/tts.py:
```diff
@@ -0,0 +1,169 @@
+
+        speed: Optional[float] = None
+        config: Optional[list[ConfigOption]] = None
+        base_url: str = "https://api.models.hathora.dev/inference/v1/tts",
```
Remove:
```suggestion
        base_url: str = "https://api.models.hathora.dev/inference/v1/tts"
```

markbackman replied:
Does this base_url work currently? I tried running the example, but I get an error:
```
Cannot connect to host api.models.hathora.dev:443 ssl:default [nodename nor servname provided, or not known]
```

markbackman commented on examples/foundational/07ag-interruptible-hathora.py:
```diff
@@ -0,0 +1,142 @@
+from dotenv import load_dotenv
+from loguru import logger
+
+from pipecat.audio.turn.smart_turn.base_smart_turn import SmartTurnParams
```
Remove unused import:
```suggestion
```

markbackman reviewed (commented):
Looking good. Just some minor comments. I'd like to be able to run the example, too, so perhaps you can help with the base URL. Maybe my API isn't provisioned for this URL (or something else!)

markbackman commented:
> @markbackman With the exceptions of the unresolved comments above, I believe I addressed the other concerns.
> 
> The example will not work as-is because the default API endpoints for base_url are not live yet. You'll need to make the below change to use the dev endpoint:
> 
> ```diff
>     stt = HathoraSTTService(
>         model="nvidia-parakeet-tdt-0.6b-v3",
> +        params=HathoraSTTService.InputParams(
> +            base_url="https://model-marketplace-api-dev.fly.dev/inference/v1/stt",
> +        ),
>     )
> 
>     tts = HathoraTTSService(
>         model="hexgrad-kokoro-82m",
> +        params=HathoraTTSService.InputParams(
> +            base_url="https://model-marketplace-api-dev.fly.dev/inference/v1/tts",
> +        ),
>     )
> ```
> 
> I moved `base_url` to the `InputParams` because it should very rarely need to be changed by end users, and would likely only be changed by Hathora devs to test new staging endpoints.

Following Pipecat's patterns, it should be in the `__init__`, even if it's rarely changed. This is where initialization time params go.

mikeseese commented:
Thanks @markbackman; I've made the requested changes (and ran the linter; I'm sorry I wasn't doing that before to catch some of the basic things 😢 ).

Are you having issues with the committed or default API endpoint? `api.models.hathora.dev` is **not live yet** and will not work as-is. You need to make a modification to the `07ag-interruptible-hathora.py` to get it to work. Here is the full file with the modified example using the dev endpoint (`model-marketplace-api-dev.fly.dev`):

[07ag-interruptible-hathora.py](https://github.com/user-attachments/files/24547155/07ag-interruptible-hathora.py)

I'm fully aware (and made sure others on our team are as well) that you can't merge this until the committed default works; I've been just trying to streamline the review process while someone else on the team is still trying to get the official `hathora.dev` subdomain to work. I'm sorry for any inconvenience or confusion this has caused.
